### PR TITLE
[MultiDB] Add multidb support to sonic-ztp

### PR DESCRIPTION
**Pull request by rajendra-dendukuri**

### CODE REVIEW

**Comment by rajendra-dendukuri** [comment]
retest buster please

File: src/usr/lib/ztp/ztp-profile.sh [hunk 1]
```diff
@@ -264,7 +256,7 @@ if [ "$CMD" = "remove" ] ; then
             fi
         else
             # Remove ZTP configuration from config-db
-            redis-cli -n 4 DEL "ZTP|mode"
+            sonic-db-cli CONFIG_DB DEL "ZTP|mode"
```
Comment (wendani) [inline src/usr/lib/ztp/ztp-profile.sh:259]: May need to direct to `/dev/null` to avoid `sonic-ztp[xxx]: 1` in syslog

Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/ztp-profile.sh]: Thanks. I fixed it.

**Comment by lguohan** [comment]
retest this please

File: src/usr/lib/ztp/plugins/configdb-json [hunk 2]
```diff
@@ -47,6 +50,11 @@ class ConfigDBJson:
             raise TypeError('Snmp provided with invalid argument type.')
         self.__input_file = input_file
 
+        self.configDB = ConfigDBConnector()
+        if self.configDB is None:
```
Comment (qiluo-msft) [inline src/usr/lib/ztp/plugins/configdb-json:54]: >self.configDB [](start = 11, length = 13)

Could you clarify when it is None?

File: src/usr/lib/ztp/plugins/configdb-json [hunk 3]
```diff
@@ -59,12 +67,71 @@ class ConfigDBJson:
...
+            data = self.configDB.get_entry("ZTP", "mode")
+            if data is not None and data.get("profile") is not None:
+                profile_status = data.get("profile")
+            else:
+                profile_status = None
+            if profile_status is not None and profile_status == 'active':
                 updateActivity('configdb-json: Removing ZTP configuation from Config DB')
                 logger.info('configdb-json: Configuration change detected. Removing ZTP configuation from Config DB.')
-                runCommand('redis-cli -n 4 DEL "ZTP|mode"', capture_stdout=False)
+                self.configDB.delete_table("ZTP")
```
Comment (qiluo-msft) [inline src/usr/lib/ztp/plugins/configdb-json:134]: >delete_table [](start = 30, length = 12)

Original code is deleting a entry, and new code is deleting a table. Is it expected?

File: tests/test_configdb-json.py [hunk 4]
```diff
@@ -170,6 +170,12 @@ def test_json_config_valid_load(self, tmpdir):
 
         cmd = "/bin/sed -i -e 's/\"hostname\": \".*\"/\"hostname\": \"something\"/' " + str(fh_after)
         rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"platform\": \".*\"/\"platform\": \"invalid1\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"mac\": \".*\"/\"mac\": \"invalid2\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"hwsku\": \".*\"/\"hwsku\": \"invalid3\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
```
Comment (qiluo-msft) [inline tests/test_configdb-json.py:178]: Could you just parse config-before.json, modify in memory and write to config-after.json?

File: tests/test_configdb-json.py [hunk 5]
```diff
@@ -218,13 +230,30 @@ def test_json_config_valid_reload(self, tmpdir):
...
         self.__write_file(str(fh_after), data)
 
         cmd = "/bin/sed -i -e 's/\"hostname\": \".*\"/\"hostname\": \"something\"/' " + str(fh_after)
         rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"platform\": \".*\"/\"platform\": \"invalid1\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"mac\": \".*\"/\"mac\": \"invalid2\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
+        cmd = "/bin/sed -i -e 's/\"hwsku\": \".*\"/\"hwsku\": \"" + hwsku + "_dup"  +"\"/' " + str(fh_after)
+        rc  = runCommand(cmd)
```
Comment (qiluo-msft) [inline tests/test_configdb-json.py:256]: The same

File: src/usr/lib/ztp/ztp-engine.py [hunk 6]
```diff
@@ -106,57 +108,123 @@ class ZTPEngine():
     ## Interfaces state table
     __intf_state = dict()
 
-    def __link_scan(self):
+    ## Redis DB connectors
+    configDB = None
+    applDB   = None
```
Comment (qiluo-msft) [inline src/usr/lib/ztp/ztp-engine.py:113]: You mixed using Class Variables and Instance Variables. Please check some tutorial, such as [https://careerkarma.com/blog/python-class-variables/](https://careerkarma.com/blog/python-class-variables/)

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comments

File: src/usr/lib/ztp/plugins/configdb-json (hunk 2 above)
Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/plugins/configdb-json:59]: Fixed in latest commit

File: src/usr/lib/ztp/plugins/configdb-json (hunk 3 above)
Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/plugins/configdb-json]: Fixed in latest commit

File: tests/test_configdb-json.py (hunk 4 above)
Comment (rajendra-dendukuri) [reply tests/test_configdb-json.py]: Fixed in latest commit

File: tests/test_configdb-json.py (hunk 5 above)
Comment (rajendra-dendukuri) [reply tests/test_configdb-json.py]: Fixed in latest commit

File: src/usr/lib/ztp/ztp-engine.py (hunk 6 above)
Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/ztp-engine.py]: Fixed in latest commit

**Comment by qiluo-msft** [comment]
@dzhangalibaba Could you also help review this PR?

File: src/usr/lib/ztp/plugins/configdb-json (hunk 2 above)
Comment (qiluo-msft) [reply src/usr/lib/ztp/plugins/configdb-json:59]: The new code seems worse. Why construct ConfigDBConnector twice?

---
In reply to: [492363061](https://github.com/Azure/sonic-ztp/pull/16#discussion_r492363061) [](ancestors = 492363061)

Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/plugins/configdb-json]: My bad. I did not resolve the merge conflict properly and ended up having two copies of the code. I fixed it now.

**Review by qiluo-msft — APPROVED** [review]
LGTM. Please also resolve other reviewers' comments.

File: src/usr/lib/ztp/ztp-engine.py [hunk 7]
```diff
@@ -79,84 +81,151 @@ class ZTPEngine():
...
-            if rc == 0 and len(op) != 0 and op[0] == 'active':
-                self.__link_scan_enabled = 'True'
-            else:
-                self.__link_scan_enabled = 'False'
-
-        if self.__link_scan_enabled == 'False':
+        # Connect to AppDB
+        try:
+            if self.applDB is None:
+                self.applDB = SonicV2Connector(host='127.0.0.1')
```
Comment (dzhangalibaba) [inline src/usr/lib/ztp/ztp-engine.py:134]: remove "host='127.0.0.1' " , it is not needed , all info should come from database_config.json and the host here didn't take effect.

**Review by dzhangalibaba — CHANGES_REQUESTED** [review]

File: src/usr/lib/ztp/ztp-engine.py (hunk 7 above)
Comment (rajendra-dendukuri) [reply src/usr/lib/ztp/ztp-engine.py]: Implemented the recommended change.

**Review by dzhangalibaba — APPROVED** [review]

**Pull request merged**